group/assignment6/TestRearrangeCars.cpp

```diff
@@ -0,0 +1,280 @@
+	vector<int> startStates = { -1 };
+	vector<int> endStates = { -1 };
+	vector<set<int>> constraints = { { 0 } };
+
+	testInvalidInput(startStates, endStates, constraints);
+}
+
+void TestRearrangeCars::testValidStatesWithEmptyConstraints() {
+	vector<int> startStates = { 0, 1, -1, 2 };
+	vector<int> endStates = { 2, -1, 1, 0 };
+	vector<set<int>> constraints = { {},{},{},{} };
+
+	testValidInput(startStates, endStates, constraints);
+}
+
+void TestRearrangeCars::testValidStatesWithInvalidConstraints1()
+{
+	vector<int> startStates = { 0, 1, -1, 2 };
+	vector<int> endStates = { 2, -1, 1, 0 };
+	vector<set<int>> constraints = { {},{ -1 },{},{} };
+
+	testInvalidInput(startStates, endStates, constraints);
+}
+
+void TestRearrangeCars::testValidStatesWithInvalidConstraints2()
+{
+	vector<int> startStates = { 0, 1, -1, 2 };
+	vector<int> endStates = { 2, -1, 1, 0 };
+	vector<set<int>> constraints = { { 0 },{ 3 },{},{} };
+
+	testInvalidInput(startStates, endStates, constraints);
+}
+
+void TestRearrangeCars::testValidStatesWithInvalidConstraints3()
+{
+	vector<int> startStates = { 0, 1, -1, 2 };
+	vector<int> endStates = { 2, -1, 1, 0 };
+	vector<set<int>> constraints = { { 0 },{ 2 },{},{} };
+
+	testInvalidInput(startStates, endStates, constraints);
+}
+
+void TestRearrangeCars::testValidStatesWithImpossibleConstraints1()
+{
+	vector<int> startStates = { 0, 1, -1, 2 };
+	vector<int> endStates = { 2, -1, 1, 0 };
+	vector<set<int>> constraints = { {},{ 2 },{ 1 },{} };
+
+	testInvalidInput(startStates, endStates, constraints);
+}
+
+void TestRearrangeCars::testValidStatesWithImpossibleConstraints2()
+{
+	vector<int> startStates = { 0, 1, -1, 2 };
+	vector<int> endStates = { 2, -1, 1, 0 };
+	vector<set<int>> constraints = { {},{ 2 },{ 0, 1, 2 },{} };
+
+	testInvalidInput(startStates, endStates, constraints);
+}
+
+void TestRearrangeCars::testValidStates1() {
+	vector<int> startStates = { 0, 1, -1, 2 };
+	vector<int> endStates = { 2, -1, 1, 0 };
+	vector<set<int>> constraints = { {},{ 2 },{},{} };
+
+	testValidInput(startStates, endStates, constraints);
+}
+
+void TestRearrangeCars::testValidStates2() {
+	vector<int> startStates = { 0, 1, -1, 2 };
+	vector<int> endStates = { 0, 2, -1, 1 };
+	vector<set<int>> constraints = { {},{},{ 0 },{} };
+
+	testValidInput(startStates, endStates, constraints);
+}
+
+void TestRearrangeCars::testValidStates3() {
+	vector<int> startStates = { 0, 1, -1, 2 };
+	vector<int> endStates = { 0, 2, -1, 1 };
+	vector<set<int>> constraints = { { 2 },{},{ 0 },{} };
+
+	testValidInput(startStates, endStates, constraints);
+}
+
+void TestRearrangeCars::testValidStatesAllSequences1() {
+	vector<int> startStates = { 0, 1, -1, 2 };
+	vector<int> endStates = { 2, -1, 1, 0 };
+
+	testValidInputAllSequences(startStates, endStates);
+}
+
+void TestRearrangeCars::testValidStatesAllSequences2() {
+	vector<int> startStates = { 0, 1, -1 };
+	vector<int> endStates = { 1, 0, -1 };
+

```

Comment:
I can try to see why it doesn't work but if it takes too much time I will remove it. Is that ok?

Reply:
Sure! Now I again think that this might be my mistake as I was not passing constraints in testValidInputAllSequences... Was in the hurry, sorry. But I will be able to check this only few hours later.